spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -261,31 +262,36 @@ protected void afterUnbindConsumer(String destination, String group, C consumerP
 
 		private final boolean extractEmbeddedHeaders;
 
-		private ReceivingHandler(boolean extractEmbeddedHeaders) {
+		private final boolean supportDeserializationNatively;
+
+		private ReceivingHandler(boolean extractEmbeddedHeaders, boolean supportDeserializationNatively) {
 			this.extractEmbeddedHeaders = extractEmbeddedHeaders;
+			this.supportDeserializationNatively = supportDeserializationNatively;
 		}
 
 		@Override
 		@SuppressWarnings("unchecked")
 		protected Object handleRequestMessage(Message<?> requestMessage) {
-			MessageValues messageValues;
-			if (this.extractEmbeddedHeaders) {
-				try {
-					messageValues = AbstractMessageChannelBinder.this.embeddedHeadersMessageConverter.extractHeaders(
+			MessageValues messageValues = (this.extractEmbeddedHeaders) ? extractEmbeddedHeaders(requestMessage) :
+					new MessageValues(requestMessage);
```

Comment:
Probably doing things in a different order helps - check if native deserialization happened (payload is not byte[]), then extract headers.